### ISSUE TITLE
type narrowed return value limitation with typevar constraints

### ISSUE THREAD

**Issue by skeggse** [microsoft/pyright#8244]
**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

Type narrowing doesn't seem to allow a subtype of the available type constraint. In particular, if you have a function that transforms the input only if it's a specific subtype, then it doesn't allow the return type in a branch that narrows the type to that subtype to match that subtype.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```py
T = TypeVar("T", str, str | None)


def transform_value(value: str) -> str:
    return value + value


def process_value(value: T) -> T:
    if value is None:
        return value
    # error: Expression of type "str" is incompatible with return type "T@process_value"
    return transform_value(value)
```

It's possible that I'm just misusing `TypeVar`s here, but it seems reasonable to me to consider `process_value` to be in one of two "modes" - either it's received something `Optional` or not. I could work around this with `@overload` but I'd rather make the function self-consistent.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

`pyright` CLI v1.1.369 and VS Code

**Comment by erictraut** [comment]
You're arguably misusing TypeVars here. Value-constrained type variables are poorly spec'ed currently, and you're in a gray area of the spec. The problem is that you've specified two constraints (`str` and `str | None`) where one of the constraints (`str`) is a proper subtype of the other constraint. The spec doesn't indicate what should happen in this case. Should it be an error to express such constraints? Probably, because the resulting behavior is ambiguous at best.

In general, I recommend against using value-constrained type variables if you can avoid it. It's a very odd construct — unique to Python. It's poorly spec'ed, and the resulting behaviors are highly dependent on internal implementation details of type checkers.

Refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables) and [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables) for insights into pyright's handling of value-constrained type variables and their limitations.

In any case, pyright is working as intended here, so I'm going to close this issue.

**Comment by skeggse** [comment]
Thanks for the explanation!
